### ISSUE TITLE
About implementation of NCE Loss

### ISSUE THREAD

**Issue by BBBBchan** [maeve07/RCA#9]
I have some confusion about the implementation of NCE loss. The main code is in `rca.py: line204-223`

In my understanding, after 
```python
x_mid_pool_norm = F.normalize(x_mid_pool.unsqueeze(0)).squeeze()
feat_neg = F.normalize(feat_memory)
```
, we have gotten the normalized $f_l$ and all $m_l$ in the memory bank, the cosine similarity can be calculated with code :
```python
similarity_neg = torch.einsum('c,nc->n', [x_mid_pool_norm, feat_neg.detach()])
logit_neg = torch.div(similarity_neg, self.temperature)
```
If so, what is the purpose of these codes?
```python
max_log = torch.max(logit_neg)
exp_logit_neg = torch.exp(logit_neg - max_log.detach())
```
It seems like this operation minus the max value of cosine similarity between $f_l$ with every $m_l$ in the memory bank before exp.
Also, while calculating the cosine similarity between $f_l$ and $m_l^{+}$, the code is still minus the `max_log` item:
```python
similarity_pos = torch.einsum('c,nc->n', [x_mid_pool_norm, feat_pos.detach()])
logit_pos = torch.div(similarity_pos, self.temperature)
logit_pos = logit_pos - max_log.detach()
exp_logit_pos = torch.exp(logit_pos)
```
What is that purpose, and which part of the formula does this code correspond to?

**Comment by tfzhou** [comment]
Hi @BBBBchan, the purpose to subtract ```max_log``` is for numerical stability.

**Comment by BBBBchan** [comment]
Thanks for the reply!
I have another problem when generating the pseudo labels. After training the model, I've got a pth file in run/pascal/model and images in run/pascal/feat. However, when I try to generate the pseudo labels referring to `genlabel.py`, I cannot find the saliency map folder and attr folder, which correspond to sal_map and att_path in `genlabel.py`. Could you please introduce how to generate the sal map and attr for each image?

Also, I refer to #5 and find that they use the attr in 'run/pascal/feat' . However, the attr files in 'run/pascal/feat' are named by indexes instead of the image names, which cannot be directly corresponded and used in `genlabel.py`.

**Comment by Shield9999** [comment]
I have same issue with you @BBBBchan.

**Comment by maeve07** [comment]
Hello, regarding attr, you can refer to the upsampling code we uploaded, and then use `genlabel.py` to generate pseudo labels. Then about the sal map you can refer to the paper.

**Comment by Shield9999** [comment]
Hello author. What upsampling code do you mean?

**Comment by wangjialipp** [comment]
作者大大，没有明白为什么你训练的时候要用train_cls.txt，为什么要加入类别
